src/policy_artifacthub.rs

```diff
@@ -0,0 +1,325 @@
+    pub url: Url,
+}
+
+fn default_recommendation() -> Url {
+    Url::parse("https://artifacthub.io/packages/helm/kubewarden/kubewarden-controller").unwrap()
+}
+
+// #[derive(Default, Deserialize, Serialize, Debug, Clone)]
+// #[serde(rename_all = "camelCase")]
+// pub struct Maintainer {
+//     name: String,
+//     email: String,
+// }
+
+#[derive(Default, Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Provider {
+    #[serde(default = "default_provider")]
+    name: String,
+}
+
+fn default_provider() -> String {
+    "kubewarden".to_string()
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Annotation {
+    /// required, derived from metadata
+    #[serde(rename = "kubewarden/mutation")]
+    KubewardenMutation(bool),
+    #[serde(rename = "kubewarden/contextAware")]
+    KubewardenContextAware(bool),
+
+    /// optional, derived from metadata.annotations
+    #[serde(rename = "kubewarden/resources")]
+    KubewardenResources(String),
+    #[serde(rename = "rancher/hidden-ui")]
+    RancherHiddenUi(bool),
+    #[serde(rename = "kubewarden/rules")]
+    KubewardenRules(String),
+    #[serde(rename = "kubewarden/questions-ui")]
+    KubewardenQuestionsUi(String),
+}
+
+impl ArtifactHubPkg {
+    pub fn from_metadata(
+        metadata: &MetadataV2,
+        version: &str,
+        readme: Option<&str>,
+        questions: Option<&str>,
+    ) -> Result<Self> {
+        // metadata annotations can't be None
+        if metadata.annotations.is_none() {
+            return Err(anyhow!(
+                "no annotations in policy metadata. policy metadata must specify annotations"
+            ));
+        }
+        // metadata annots can't be empty
+        let metadata_annots = metadata.annotations.as_ref().unwrap();
+        if metadata_annots.is_empty() {
+            return Err(anyhow!(
+                "no annotations in policy metadata. policy metadata must specify annotations"
+            ));
+        }
+
+        // build minimal struct with required fields, besides those derived from
+        // metadata.annotations:
+        let mut artifacthubpkg = ArtifactHubPkg {
+            version: Version::parse(version).expect("policy version must be in semver"),
+            created_at: OffsetDateTime::now_utc(),
+            provider: Provider {
+                name: default_provider(),
+            },
+            recommendations: vec![Recommendation {
+                url: default_recommendation(),
+            }],
+            annotations: HashSet::from([
+                Annotation::KubewardenMutation(metadata.mutating),
+                Annotation::KubewardenContextAware(metadata.context_aware),
+                Annotation::KubewardenRules(
+                    serde_json::to_string(&metadata.rules).expect("policy rules key is malformed"),
+                ),
+            ]),
+            // the following required fields will be filled later on:
```

Comment:
this is the incredibly ugly part.
it could be improved by setting those required fields right here and then. For example, for the `artifacthubpkg.name` field:

```rust
            name: {
                let mut missing_title = false;
                for a in metadata_annots { // could be 0 loops, but we checked is not empty above..
                   if let policy_metadata::Annotation::PolicyTitle(ref title) = a {
                       // found
                       title
                    } else {
                        // not found
                    }
                    if missing_title {
                        return Err(anyhow!(
                            "policy metadata must specify {} in annotations",
                            "\"io.kubewarden.policy.title\"",
                        ));
                    }
                }
            },
```

Yet something like this is not easily parametrized to transform as function. As we need to do this for `String` and `Url` (`Link`)

Reply:
That, or the loop you wrote below. I've just seen that

Reply:
Resolving, as this is irrelevant now.